Calculate a photoacclimation mixed layer depth in MOM6

theresa-cordero opened the pull request:
This PR changes how the photoacclimation depth is calculated, allowing MOM6  to use it's mixed layer depth diagnosic routines to calculate the depth and pass that two dimensional field to COBALTv3. 

These changes must be done afte MOM_diagnose_MLD.F90 is merged into the dev/cefi branch ([PR 15](https://github.com/NOAA-CEFI-Regional-Ocean-Modeling/MOM6/pull/15)  or from dev/gfdl) and the generic tracer interfaces are changed ([ocean_BGC PR 88](https://github.com/NOAA-CEFI-Regional-Ocean-Modeling/ocean_BGC/pull/88) ).

These changes introduce additional flexibility in how the photoacclimation depth is calculated at runtime through a number of new parameters.

- `PHA_MLD_CALC` (default false) if false, use a fixed value for the photoacclimation mixed layer depth. This option should be used to skip calculating a MLD. Likely this will be true for people who are not running COBLATv3 and do not want to calculate an unused depth.
- `PHA_MLD_VAL` (default 0.0) This is the fixed value for `PHA_MLD_CALC`. The purpose of using 0.0 as a default is to lead to a crash or obviously wrong solution when using COBALTv3. A warning message has been added to COBALT to give users some insight into the problem.  
- `PHA_MLD_USE_DELTA_RHO` (default false) If true, use a density difference to calculate the photoacclimation MLD. 
- `PHA_MLD_HREF` (default 0.0) Similar to the diagnostic MLD, the reference depth can be adjusted. A value of 10.0m would be most similar to de Boyer Montegut et al. 2004
- `PHA_MLD_DRHO` (default 0.03) The target difference in the density at h_ref and the bottom of photoacclimation MLD. 
- `PHA_MLD_USE_DELTA_ENG` (default false) If true, use the energy based MLD to find the photoacclimation depth (not tested).
- `PHA_MLD_DENG` (default 25.0) The threshold for the energy based MLD (not tested)
- `PHA_MLD_TIMESCALE` (default 3 days, unused) An additional argument to do time averaging of the photoacclimation depth. I thought this was done in COBALT, but it doesn't seem like it. We may want this feature when the `PHA_MLD_USE_DELTA_ENG` is true so I have left it included for now. 

There are some differences in how `diagnoseMLDbyDensityDifference` calculated the MLD and what was previously used in COBALT. Therefore, we expect some minor changes in the solutions.... [results of the NWA12 5 year run here]

Commit:
Calculate a photoacclimation MLD in MOM6

Using the diagnose MLD routines calculate a MLD that can be used
in COBALTv3.

theresa-cordero commented:
note: after the mixed layer depth diagnostic routines are updated, I will need to check that the interfaces in this PR are correct.

yichengt900 commented on src/tracer/MOM_generic_tracer.F90:
```diff
@@ -572,6 +614,16 @@ subroutine MOM_generic_tracer_column_physics(h_old, h_new, ea, eb, fluxes, Hml,
+      mld_pha(:,:) = CS%mld_pha_val
+    else
+      if (CS%mld_pha_use_delta_rho) then
+        call diagnoseMLDbyDensityDifference(-1, h_old, tv, CS%mld_pha_drho, G, GV, US, CS%diag, ref_h_mld=CS%mld_pha_href, MLD_out=mld_pha)
```
This needs to be updated:

`        call diagnoseMLDbyDensityDifference(-1, h_old, tv, CS%mld_pha_drho, G, GV, US, CS%diag, CS%mld_pha_href, -1, -1, MLD_out=mld_pha)`

Also `MLD_out`is msssing in current `diagnoseMLDbyDensityDifference` and `diagnoseMLDbyEnergy` subroutines.

theresa-cordero replied:
This should be added to the diagnose MLD routines with this PR.

yichengt900 commented on src/tracer/MOM_generic_tracer.F90:
```diff
@@ -582,7 +634,8 @@ subroutine MOM_generic_tracer_column_physics(h_old, h_new, ea, eb, fluxes, Hml,
                G%areaT, get_diag_time_end(CS%diag), &
                optics%nbands, optics%max_wavelength_band, optics%sw_pen_band, optics%opacity_band, &
-               internal_heat=tv%internal_heat, frunoff=fluxes%frunoff, sosga=sosga, geolat=G%geolatT, eqn_of_state=tv%eqn_of_state)
+               internal_heat=tv%internal_heat, frunoff=fluxes%frunoff, sosga=sosga, geolat=G%geolatT, eqn_of_state=tv%eqn_of_state, &
```
We can consider remove `eqn_of_state=tv%eqn_of_state`

theresa-cordero replied:
Agreed - once we have decided that these results are similar enough to those using the photoacclimation depth calculated in COBALT we can change this interface.

theresa-cordero replied:
I think once this PR and the ocean_bgc PR are merged, I will open two more which remove the extra arguments.

yichengt900 commented on src/tracer/MOM_generic_tracer.F90:
```diff
@@ -593,14 +646,16 @@ subroutine MOM_generic_tracer_column_physics(h_old, h_new, ea, eb, fluxes, Hml,
                opacity_band=G%US%m_to_Z*optics%opacity_band(:,:,:,:), &
                internal_heat=G%US%RZ_to_kg_m2*US%C_to_degC*tv%internal_heat(:,:), &
-               frunoff=G%US%RZ_T_to_kg_m2s*fluxes%frunoff(:,:), sosga=sosga, geolat=G%geolatT, eqn_of_state=tv%eqn_of_state)
+               frunoff=G%US%RZ_T_to_kg_m2s*fluxes%frunoff(:,:), sosga=sosga, geolat=G%geolatT, eqn_of_state=tv%eqn_of_state, &
```
Same here

yichengt900 commented on src/tracer/MOM_generic_tracer.F90:
```diff
@@ -593,14 +646,16 @@ subroutine MOM_generic_tracer_column_physics(h_old, h_new, ea, eb, fluxes, Hml,
                sw_pen_band=G%US%QRZ_T_to_W_m2*optics%sw_pen_band(:,:,:), &
                opacity_band=G%US%m_to_Z*optics%opacity_band(:,:,:,:), &
-               frunoff=G%US%RZ_T_to_kg_m2s*fluxes%frunoff(:,:), sosga=sosga, geolat=G%geolatT, eqn_of_state=tv%eqn_of_state)
+               frunoff=G%US%RZ_T_to_kg_m2s*fluxes%frunoff(:,:), sosga=sosga, geolat=G%geolatT, eqn_of_state=tv%eqn_of_state, &
```
Same.

yichengt900 reviewed (commented):
@theresa-morrison, thanks for opening this PR. The preliminary results can be found [here.](https://github.com/NOAA-CEFI-Regional-Ocean-Modeling/ocean_BGC/pull/88#issuecomment-2269466240) They are very promising, and I have just left a few comments for now.

Commit:
Update interfaces for diagnoseMLD

Some previously optional arguments were made nonoptinal in the
final version merged into dev/gfdl. So these areguents have been
added to the call in MOM_generic_tracers.

Commit:
Add unset variable in diagnoseMLDbyenergy

Commit:
Remove timescale variable

This parameter seems unneeded at this time.

theresa-cordero commented:
I think I addressed the interface updates with the diagnose MLD routines. This PR should be ready to go.

yichengt900 commented on src/tracer/MOM_generic_tracer.F90:
```diff
@@ -423,6 +433,33 @@ subroutine initialize_MOM_generic_tracer(restart, day, G, GV, US, h, tv, param_f
+                 "If true, use a fixed value for the photoacclimation mixed layer depth within the "//&
+                 "generic tracer update. This MLD is only used for photoacclimation.", default=.true.)
+    if (CS%mld_pha_fixed) then
+      call get_param(param_file, "MOM", "PHA_MLD_VAL", CS%mld_pha_val, &
```
Would it be more appropriate to set the default `mld_pha_val` to 75 meters to better represent a global average MLD when `PHA_MLD_FIXED` is true? @charliestock, do you have any additional thoughts on this?

theresa-cordero replied:
We can change it to a more realistic value, 75m or whatever @charliestock recommends.

theresa-cordero replied:
@charliestock should chime in here. 
We discussed it last week and the choice of a "bad" default is because a constant value should not be used with COBALTv3. However, we need some value in the feild for non-COBALTv3 simulations that still se the generic tracer interface.

yichengt900 commented on src/tracer/MOM_generic_tracer.F90:
```diff
@@ -507,6 +544,7 @@ subroutine MOM_generic_tracer_column_physics(h_old, h_new, ea, eb, fluxes, Hml,
     real, dimension(SZI_(G),SZJ_(G),SZK_(GV)) :: rho_dzt ! Layer mass per unit area [kg m-2]
     real, dimension(SZI_(G),SZJ_(G),SZK_(GV)) :: dzt     ! Layer vertical extents [m]
     real, dimension(SZI_(G),SZJ_(G),SZK_(GV)) :: h_work  ! A work array of thicknesses [H ~> m or kg m-2]
+    real, dimension(SZI_(G),SZJ_(G)) :: mld_pha          ! 
```
We haven't initialized a specific value for `mld_pha`, but it shouldn't be an issue since it will be assigned a new value based on the user's options. However, would it be safer to initialize it to 0 just to be extra cautious? It might also be a good idea to add a short description for this new variable.

theresa-cordero replied:
I added a description and an initial value.

Commit:
Change paramter name and though up PR

Change "MLD_PHA_FIXED" to "MLD_PHA_CALC" which is default false.
Touch up a few formatting issues.
Add initialization and comment for mld_pha.

yichengt900 commented on src/tracer/MOM_generic_tracer.F90:
```diff
@@ -572,6 +611,17 @@ subroutine MOM_generic_tracer_column_physics(h_old, h_new, ea, eb, fluxes, Hml,
+        call diagnoseMLDbyDensityDifference(-1, h_old, tv, CS%mld_pha_drho, G, GV, US, CS%diag, CS%mld_pha_href, id_ref_z=-1, id_ref_rho=-1, MLD_out=mld_pha)
+      elseif (CS%mld_pha_use_delta_eng) then
+        call diagnoseMLDbyEnergy((/-1, -1, -1/), h_old, tv, G, GV, US, (/CS%mld_pha_deng, CS%mld_pha_deng, CS%mld_pha_deng/), CS%diag, MLD_out=mld_pha)
+      endif
```
@theresa-morrison, do you think it would be better to add a warning message here in case users accidentally set both `CS%mld_pha_use_delta_rho` and `CS%mld_pha_use_delta_eng` while still having `CS%mld_pha_calc` turned on? I believe our `all_same` check should catch this at bgc side, but providing a more informative message would clarify that not only is `CS%mld_pha_calc` turned on, but either `CS%mld_pha_use_delta_rho` or `CS%mld_pha_use_delta_eng` needs to be enabled as well in order to get spatial varying `mld_pha`.

theresa-cordero replied:
yeah, that is a good idea. 
If mld_pha_calc is false, the settings of mld_pha_use_delta_rho and mld_pha_use_delta_eng are actually not read at all. I will modify the logic around where the parameters are read to give a warning to keep things simple here.

Commit:
Add fatal errors

Add fatal errors if PHA_MLD_CALC is true, but a method is not specified.

theresa-cordero commented on src/tracer/MOM_generic_tracer.F90:
```diff
@@ -423,6 +432,42 @@ subroutine initialize_MOM_generic_tracer(restart, day, G, GV, US, h, tv, param_f
+                   "If true, use an energy difference to find the photoacclimation mixed layer depth "//&
+                   "with the generic tracer update. This MLD is only used for photoacclimation.", default=.false.)
+      if (CS%mld_pha_use_delta_rho .and. CS%mld_pha_use_delta_eng) then
+        call MOM_error(FATAL, "PHA_MLD_CALC is set to true and PHA_MLD_USE_DELTA_RHO and PHA_MLD_USE_DELTA_ENG "// &
```
These error messages force users to specify an option for calculating the PHA_MLD

yichengt900 commented:
@theresa-morrison, thanks for the new updates! Everything looks good to me. I just have one additional comment:

We should probably apply the same new interface change in `config_src/external/GFDL_ocean_BGC/generic_tracer.F90` as well, just in case someone wants to compile a physics-only model (in case no external ocean_BGC repo is involved):

https://github.com/NOAA-CEFI-Regional-Ocean-Modeling/MOM6/blob/e04b63706af1828df3497110b9df4275998b1f8b/config_src/external/GFDL_ocean_BGC/generic_tracer.F90#L73

Commit:
Add additional argument to generic tracer stub

Add the photo_acc_mld argument to the generic tracer stub in MOM6.

theresa-cordero commented:
Good point. Done!

yichengt900 reviewed (approved):
All changes look good now. We will revisit the `eqn_of_state argument`, as well as explore a better way to handle all the optional arguments when we're satisfied with the results and ready to push these changes back to the `dev/gfdl` main repo. Thanks again for this great work, @theresa-morrison !
